File: leetcode/15_3sum.py
class Solution:
    def threeSum(self, nums):
        triplets = []

        if len(nums) < 3: return []

        nums.sort()
        hashes = {num:index for index, num in enumerate(nums)}

        for i in range(len(nums) - 2):
            if nums[i] > 0 : break

            for j in range(i+1, len(nums) - 1):
                k_num = -1*(nums[i] + nums[j])
                k = hashes.get(k_num)

                if k and k > j:
                    _sort = [nums[i], nums[j], nums[k]].sort()
                    triplets.append(_sort)

        return triplets


nums = [3,0,-2,-1,1,2]
s = Solution().threeSum(nums)
print(s)

File: leetcode/15_3sum.js
var threeSum = function(nums) {
    nums.sort((a, b) => a - b); // Sorted Array
    let answer = [];
    
    if (nums.length < 3) {
        return answer;
    }
    
    if (nums[0] > 0) {
        return answer;
    }
    
    let hashMap = new Map();
    
    for (let i = 0; i < nums.length; ++i) {
        hashMap.set(nums[i], i);
    }
    
    for (let i = 0; i < nums.length - 2; ++i) {
        if (nums[i] > 0) {
            break;
        }
        
        for (let j = i + 1; j < nums.length - 1; ++j) {
            let required = -1 * (nums[i] + nums[j]);
            if (hashMap.has(required) && hashMap.get(required) > j) {
                answer.push([nums[i], nums[j], required]);
            }
            j = hashMap.get(nums[j]);
        }
        
        i = hashMap.get(nums[i]);
    }
    
    return answer;
};

nums = [3,0,-2,-1,1,2]
nums = [-1,0,1,2,-1,-4]

data = threeSum(nums)
console.log(data)
